contracts/staking/GoodBatStaking.sol

```diff
@@ -0,0 +1,152 @@
+	 * @param _batUsdOracleAddress Address of the BAT/USD price oracle
+	 */
+	constructor(
+		address _token,
+		address _iToken,
+		uint256 _blockInterval,
+		NameService _ns,
+		string memory _tokenName,
+		string memory _tokenSymbol,
+		uint64 _maxRewardThreshold,
+		address _batUsdOracleAddress
+	)
+		SimpleStaking(
+			_token,
+			_iToken,
+			_blockInterval,
+			_ns,
+			_tokenName,
+			_tokenSymbol,
+			_maxRewardThreshold
+		)
+	{
+		batUsdOracleAddress = _batUsdOracleAddress;
+	}
+
+	address public batUsdOracleAddress;
+
+	/**
+	 * @dev stake some BAT
+	 * @param _amount of BAT to stake
+	 */
+	function mintInterestToken(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		uint256 res = cToken.mint(_amount);
+
+		if (
+			res > 0
+		) //cDAI returns >0 if error happened while minting. make sure no errors, if error return DAI funds
+		{
+			require(res == 0, "Minting cBat failed, funds returned");
+		}
+	}
+
+	/**
+	 * @dev redeem BAT from compound
+	 * @param _amount of BAT to redeem
+	 */
+	function redeem(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		require(cToken.redeemUnderlying(_amount) == 0, "Failed to redeem cBat");
+	}
+
+	function redeemUnderlying(uint256 _amount)
+		internal
+		override
+		returns (address, uint256)
+	{
+		uint256 tokenBalance = token.balanceOf(address(this));
+		cERC20 cToken = cERC20(address(iToken));
+		require(cToken.redeem(_amount) == 0, "Failed to redeem cBat");
+		uint256 redeemedAmount = token.balanceOf(address(this)) - tokenBalance;
+		address daiAddress = nameService.getAddress("DAI");
+		address[] memory path = new address[](2);
+		path[0] = address(token);
+		path[1] = daiAddress;
+		Uniswap uniswapContract =
+			Uniswap(nameService.getAddress("UNISWAP_ROUTER"));
+		token.approve(address(uniswapContract), redeemedAmount);
+		uint256[] memory swap =
+			uniswapContract.swapExactTokensForTokens(
+				redeemedAmount,
+				0,
+				path,
+				address(this),
+				block.timestamp
+			);
+
+		uint256 dai = swap[1];
+		require(dai > 0, "token selling failed");
+		return (daiAddress, swap[1]);
+	}
+
+	/**
+	 * @dev returns Bat to cBat Exchange rate.
+	 */
+	function exchangeRate() internal view override returns (uint256) {
+		cERC20 cToken = cERC20(address(iToken));
+		return cToken.exchangeRateStored();
+	}
+
+	/**
+	 * @dev returns decimals of token.
+	 */
+	function tokenDecimal() internal view override returns (uint256) {
+		ERC20 token = ERC20(address(token));
+		return uint256(token.decimals());
+	}
+
+	/**
+	 * @dev returns decimals of interest token.
+	 */
+	function iTokenDecimal() internal view override returns (uint256) {
+		ERC20 cToken = ERC20(address(iToken));
+		return uint256(cToken.decimals());
+	}
+
+	/**
+	 @dev function calculate BAT price in USD 
+	 @dev _amount Amount of BAT to calculate worth of it
+	 @return Returns worth of BATs in USD
+	 */
+	function getTokenPriceInUSD(uint256 _amount)
```

Comment:
we said this will be a helper function in simplestaking that developer can override, using uniswap by default
could be also chainlink by default (but would require passing the oracle address)